File: benchmarks/cross-language/go/ring.go
package main

import (
	"fmt"
	"sync/atomic"
	"time"
)

const RING_SIZE = 100
const ROUNDS = 100000

type RingActor struct {
	value    int64
	received int64
}

func main() {
	fmt.Println("=== Ring Benchmark (Go) ===")
	fmt.Printf("Ring size: %d actors\n", RING_SIZE)
	fmt.Printf("Rounds: %d\n\n", ROUNDS)

	actors := make([]RingActor, RING_SIZE)

	start := time.Now()

	// Initialize
	atomic.StoreInt64(&actors[0].value, 1)
	atomic.StoreInt64(&actors[0].received, 1)

	currentPos := 0
	for round := 0; round < ROUNDS; round++ {
		for i := 0; i < RING_SIZE; i++ {
			next := (currentPos + 1) % RING_SIZE
			val := atomic.LoadInt64(&actors[currentPos].value)
			atomic.StoreInt64(&actors[next].value, val+1)
			atomic.AddInt64(&actors[next].received, 1)
			currentPos = next
		}
	}

	elapsed := time.Since(start)

	totalMessages := ROUNDS * RING_SIZE
	totalCycles := float64(elapsed.Nanoseconds()) * 3.0
	cyclesPerMsg := totalCycles / float64(totalMessages)
	msgPerSec := float64(totalMessages) / elapsed.Seconds()

	fmt.Printf("Total messages: %d\n", totalMessages)
	fmt.Printf("Cycles/msg: %.2f\n", cyclesPerMsg)
	fmt.Printf("Throughput: %.0f M msg/sec\n", msgPerSec/1000000)
}

File: benchmarks/cross-language/rust/src/ring.rs
use std::sync::atomic::{AtomicI64, Ordering};
use std::time::Instant;

const RING_SIZE: usize = 100;
const ROUNDS: i64 = 100_000;

struct RingActor {
    value: AtomicI64,
}

fn main() {
    println!("=== Rust Ring Benchmark ===");
    println!("Ring size: {} actors", RING_SIZE);
    println!("Rounds: {}\n", ROUNDS);
    
    let actors: Vec<RingActor> = (0..RING_SIZE)
        .map(|_| RingActor {
            value: AtomicI64::new(0),
        })
        .collect();
    
    let start = Instant::now();
    
    actors[0].value.store(1, Ordering::Relaxed);
    
    let mut current = 0;
    for _ in 0..ROUNDS {
        for _ in 0..RING_SIZE {
            let next = (current + 1) % RING_SIZE;
            let val = actors[current].value.load(Ordering::Acquire);
            actors[next].value.store(val + 1, Ordering::Release);
            current = next;
        }
    }
    
    let elapsed = start.elapsed();
    
    let total_messages = ROUNDS * RING_SIZE as i64;
    let cycles = (elapsed.as_secs_f64() * 3e9) as u64;
    let cycles_per_msg = cycles as f64 / total_messages as f64;
    let throughput = 3000.0 / cycles_per_msg;
    
    println!("Total messages: {}", total_messages);
    println!("Cycles/msg: {:.2}", cycles_per_msg);
    println!("Throughput: {} M msg/sec", throughput as i64);
}
